pkg/client/client.go

```diff
@@ -0,0 +1,232 @@
+// Package client contains functions for communicating with the API server
+// Copyright (C) 2020 by Red Hat, Inc.
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+)
+
+// Request handles sending the request, handling errors, returning the response
+// socket is the path to a Unix Domain socket
+// path is the full URL path, including query strings
+// body is the data to send with POST
+// headers is a map of header:value to add to the request
+//
+// If it is successful a http.Response will be returned. If there is an error, the response will be
+// nil and error will be returned.
+func Request(socket, method, path, body string, headers map[string]string) (*http.Response, error) {
+	client := http.Client{
+		// TODO This may be too short/simple for downloading images
+		Timeout: 60 * time.Second,
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", socket)
+			},
+		},
+	}
+
+	req, err := http.NewRequest(method, "http://localhost"+path, bytes.NewReader([]byte(body)))
+	if err != nil {
+		return nil, err
+	}
+
+	for h, v := range headers {
+		req.Header.Set(h, v)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// APIErrorMsg is an individual API error with an ID and a message string
+type APIErrorMsg struct {
+	ID  string `json:"id"`
+	Msg string `json:"msg"`
+}
+
+// String returns the error id and message as a string
+func (r *APIErrorMsg) String() string {
+	return fmt.Sprintf("%s: %s", r.ID, r.Msg)
+}
+
+// APIResponse is returned with status code 400 and may contain a list of errors
+// If Status is true the Errors list will not be included or will be empty.
+// When Status is false it will include at least one APIErrorMsg with details about the error.
+// It also implements the error interface so that it can be used in place of error
+type APIResponse struct {
```

Comment:
I think the `APIReponse` structure is confusing. It's fulfilling the `error` interface but if `response.Status == true` then it's not an error.

I think we should stick to idiomatic as Go as possible. If a function can return an error, it should be the last return value in the form of `error` interface. If no error happened, that value should be nil. Variables with name `err` should always be typed as `error` and not as a type implementing `error`.

Imho all the functions which currently have `*APIResponse` struct as the last return type should return `error` instead. We should have several different Error structs denoting where the error happened (e.g. StatusCodeError, TransportError, JSONError).

What do you think? I'm very open to a discussion, this is imo a really interesting topic.

Reply:
I'm really not sure there is a right way to handle this. We have (at least) 2 completely different kinds of errors. API 'errors' aren't realy errors in the go sense, they are error responses from the API call which can in some cases have multiple errors included.

I did conflate things a bit though by converting go errors to 'ClientError' to make handling things a bit less complicated.

I don't want to add a bunch of error structs that have to be kept in-sync with the API errors, that's why they include an error ID and a message, so that the client doesn't need to know exactly what could fail.

I could split things out though and return go errors as a separate 'error' from the API response.